File: QueryBuilder.php
<?php
/**
 * The purpose of this class is to simply allow for a fairly
 * simplistic an lightweight query building tool that allows
 * one to develop a reasonably simplistic query. Please take
 * note that this class is not made to replace the likes of
 * PDO, it's simply a way in which you can write your query
 * using an object and methods attatched to this object.
 *
 * @author    Joseph Evans <joeevs196@gmail.com>
 * @version   1.0.2
 * @since     05/09/2018
 * @license   MIT
 * @copyright (c) Joseph Evans 2018
 * @todo      Review this source code, I think that it could 
 *            possibly be implemented in a better way, possibly 
 *            allowing a developer to write some sub query, 
 *            maybe think about including a 'with' method so a
 *            developer would be more than able to write some 
 *            with clause(s), etc.... 
 */

class QueryBuilder {
  private $__select;
  private $__from;
  private $__where;
  private $__joins;
  private $__joinType;
  private $__order;
  private $__orderCol;
  private $__group;

  /**
   * This is just a general constructor, nothing
   * special here.
   */
  public function __construct () {
    $this->__select = "";
    $this->__from = "";
    $this->__where = "";
    $this->__joinType = "";
    $this->__orderCol = "";
    $this->__order = "";
    $this->__group = "";
    $this->__joins = array();
  }

  /**
   * The purpose of this method is to allow a developer to
   * state the select clause within their query.
   *
   * @param  $select {string}
   * @return QueryBuilder
   */
  public function select (string $select) {
    $this->__select = $this->process($select, 'SELECT');
    return $this;
  }

  /**
   * The purpose of this method is to allow a developer to
   * state the from caluse within their query.
   *
   * @param  $from {string}
   * @return QueryBuilder
   */
  public function from (string $from) {
    $this->__from = $this->process($from, 'FROM');
    return $this;
  }

  /**
   * The purpose of this method is to allow a developer to
   * state the where clause within their query.
   *
   * @param  $where {string}
   * @return QueryBuilder
   */
  public function where (string $where) {
    $this->__where = $this->process($where, 'WHERE');
    return $this;
  }

  /**
   * The purpose of this method is to essentially allow a
   * developer to define a plain join within the query.
   *
   * @param  $table {string}
   * @return QueryBuilder
   */
  public function join (string $table) {
    if (!(isset($table) && $table != "")) {
      return $this;
    }

    $this->__joinType = $this->process($table, 'JOIN');
    return $this;
  }

  /**
   * The purpose of this method is to essentially allow a
   * developer to define an inner join within the query.
   *
   * @param  $table {string}
   * @return QueryBuilder
   */
  public function innerJoin (string $table) {
    if (!(isset($table) && $table != "")) {
      return $this;
    }

    $this->__joinType = $this->process($table, 'INNER JOIN');
    return $this;
  }

  /**
   * The purpose of this method is to essentially allow a
   * developer to define a right join within the query.
   *
   * @param  $table {string}
   * @return QueryBuilder
   */
  public function rightJoin (string $table) {
    if (!(isset($table) && $table != "")) {
      return $this;
    }

    $this->__joinType = $this->process($table, 'RIGHT JOIN');
    return $this;
  }

  /**
   * The purpose of this method is to essentially allow a
   * developer to define a left join within the query.
   *
   * @param  $table {string}
   * @return QueryBuilder
   */
  public function leftJoin (string $table) {
    if (!(isset($table) && $table != "")) {
      return $this;
    }

    $this->__joinType = $this->process($table, 'LEFT JOIN');
    return $this;
  }

  /**
   * The purpose of this method is to essentially allow
   * a developer to define what they'd like to set their
   * selected join to actually join upon.
   *
   * @param  $valueLeft   {string}
   * @param  $valueRight  {string}
   * @return QueryBuilder
   */
  public function on (string $valueLeft, string $valueRight) {
    if (!(isset($this->__joinType) && $this->__joinType != "")) {
      return $this;
    }

    if (strpos(strtoupper($this->__joinType), 'ON') === false) {
      $this->__joinType .= " ON ";
    }

    $this->__joinType .= $valueLeft . " = ". $valueRight;
    array_push($this->__joins, $this->__joinType);
    $this->__joinType = ""; // reset
    return $this;
  }

  /**
   * The purpose of this method is to simply allow
   * a developer to execute the group by clause
   * within their query
   *
   * @param  $group {string}
   * @return QueryBuilder
   */
  public function groupBy (string $group) {
    if (!(isset($group) && $group != "")) {
      return $this;
    }

    $this->__group = $this->process($group, 'GROUP BY');
    return $this;
  }

  /**
   * The purpose of this method is to simply allow
   * a developer to state what they'd like
   * to order the data by.
   *
   * @param  $args {string}
   * @return QueryBuilder
   */
  public function orderBy (string $args) {
    $column;
    $order;
    $numberOfArgs = func_num_args();

    if ($numberOfArgs > 0) { $column = func_get_arg(0); }
    if ($numberOfArgs > 1) { $order = func_get_arg(1); }

    if (!(isset($column) && $column != "")) {
      return $this;
    } else if (!isset($order) || $order == "") {
      $order = 'ASC'; // assume
    } else {
      $uOrder = strtoupper($order);
      $validOrder = ($uOrder == 'ASC' || $uOrder == 'DESC');
      if (!$validOrder) { $order = 'ASC'; } // assume
    }

    $this->__orderCol = $this->process($column, 'ORDER BY') . " ";
    $this->__order = strtoupper($order);
    return $this;
  }

  /**
   * The purpose of this method is to simply process
   * and concat some string onto another, as this
   * will be run multiple times, may as well use
   * one method to do this.
   *
   * @param $operation {string}
   * @param $reference {string}
   * @return string
   */
  private function process (string $reference, string $operation) {
    $validRef = (isset($reference) && $reference != "");
    $validOp = (isset($operation) && $operation != "");

    if ($validRef && $validOp) {
      $upper = strtoupper($operation);

      if (strpos($reference, $upper) === false) {
        $reference = " " . $upper . " " . $reference . " ";
      }
    }

    return $reference;
  }

  /**
   * The purpose of this method is to
   * simply allow a developer to see
   * the current join.
   *
   * @return string
   */
  public function getCurrentJoin () {
    return $this->__joinType;
  }

  /**
   * The purpose of this method is to
   * allow the developer to specify a
   * more complex sub query when using
   * a join.
   *
   * @param $join {string}
   * @return QueryBuilder
   */
  public function setCurrentJoin (string $join) {
    $validJoin = false;
    $completeJoin = false;

    try {
      $caps = strtoupper($join);
      $validJoin = (strpos($caps, 'JOIN') === true);
    } catch (Exception $e) {
      // don't worry about it
    }

    try {
      $caps = strtoupper(join);
      $completeJoin = (strpos($caps, 'ON') === true) && (strpos($caps, '=') === true);
    } catch (Exception $e) {
      // don't worry about it
    }

    if (isset($join) && $join != "" && $validJoin === true) {
      $this->__joinType = $join;
    }

    if ($completeJoin && $validJoin) {
      array_push($this->__joins, $this->__joinType);
      $this->__joinType = "";
    }

    return $this;
  }

  /**
   * The purpose of this method is to essentially return
   * the query string.
   *
   * @return string
   */
  public function build () {
    $str = $this->__select;
    $str .= $this->__from;
    foreach ($this->__joins as $join) {
        $str .= $join;
    }
    $str .= $this->__where;
    $str .= $this->__group;
    $str .= $this->__orderCol . $this->__order;
    return $str;
  }
}
?>

File: QueryBuilder.java

/**
 * @author  JoesephEvans <p><br/><b>Contact: </b> joeevs196@gmail.com</p>
 * @version 1.0.0
 * @class   The purpose of this class is to implement a 
 *          feature much like jOOQ, only a much more 
 *          minimal solution. This implementation is more 
 *          than likely vastly simplistic in comparison to 
 *          jOOQ.
 * 
 * @todo    Consider implementing a feature where one is able 
 *          to include many subqueries, with clause(es), etc. 
 * @todo    A simple 'switch' feature to toggle whether the 
 *          developer would like to sort in ascending order 
 *          by default or descending order by default. 
 * 
 * @note    The purpose of this code is not to replace tools 
 *          such as HQL, but rather to work along side, specifically 
 *          when using a native query. 
 */

import java.util.List;
import java.util.ArrayList;

public class QueryBuilder {
    private String select;
    private String from;
    private String where;
    private String join;
    private String orderBy;
    private String groupBy;
    private String build;
    private final List<String> joins = new ArrayList<>();
    private final static String NEW_LINE = "\n"; // short for newline 
    
    
    
    /*
     * These properties are not YET being used 
     * however the plan is to implement their use(s) at a 
     * later date. 
     */
    @SuppressWarnings("unused")
    private String with;
    private final List<String> withClauses = new ArrayList<>();
    private String subQuery;
    private String limit;
    private String offset;
    
    
    
    /**
     * This is just a very basic constructor, it serves nothing 
     * more than allowing other objects/classes the ability to 
     * create an object from this class. 
     */
    public QueryBuilder () {
        // TODO ... 
        this.build = "";
    }
    
    
    
    /**
     * @ignore 
     * @param string {String}
     * @return Boolean 
     * 
     * The purpose of this method is to simply see if 
     * a provided string is suitable to use. Don't worry, 
     * this code uses a technique called 'logical short-
     * circuiting', as you can see here, I'm checking that 
     * if the string is null first. As we check if an argument 
     * is null FIRST, this prevents a null pointer exception 
     * from occuring when we run the 'isEmpty' method on the 
     * provided string. 
     */
    private Boolean isValidString (String string) {
        return string != null && !string.isEmpty();
    }
    
    
    
    /**
     * @ignore
     * @param reference {String}
     * @param operation {String}
     * @return String
     * 
     * The purpose behind this method is to simply allow 
     * a developer the ability to combine two strings 
     * together, provided both strings are okay to use.
     */
    private String process (String reference, String operation) {
        if (isValidString(reference) && isValidString(operation)) {
            if (!reference.toUpperCase().contains(operation.toUpperCase())) {
                reference = " " + operation.toUpperCase() + " " + reference;
            }
        }
        
        return reference;
    }
    
    
    
    /**
     * @param  select {String}
     * @return QueryBuilder
     * 
     * The purpose behind this method is to allow a developer 
     * the ability to execute the 'select' statement.
     */
    public QueryBuilder select (String select) {
        this.select = process(select, "SELECT");
        return this;
    }
    
    
    
    /**
     * @param  from {String}
     * @return QueryBuilder
     * 
     * The purpose of this method is to allow a developer to include
     * a from clause within the current query.
     */
    public QueryBuilder from (String from) {
        this.from = process(from, "FROM");
        return this;
    }    
    
    
    
    /**
     * @param  where {String}
     * @return QueryBuilder 
     * 
     * The purpose of this method is to allow a developer to state 
     * the with clause.
     */
    public QueryBuilder where (String where) {
        this.where = process(where, "WHERE");
        return this;
    }
    
    
    
    /**
     * @param  join {String}
     * @return QueryBuilder
     * 
     * The purpose of this method is to allow a developer to do 
     * a simple join. 
     */
    public QueryBuilder join (String join) {
        this.join = process(join, "JOIN");
        return this;
    }
    
    
    
    /**
     * @param  join {String}
     * @return QueryBuilder 
     * 
     * The purpose of this method is to allow a developer to 
     * do a simple inner join. 
     */
    public QueryBuilder innerJoin (String join) {
        this.join = process(join, "INNER JOIN");
        return this;
    }
    
    
    
    /**
     * @param  join {String}
     * @return QueryBuilder
     * 
     * The purpose of this method is to allow a developer to 
     * perform a simple right join.
     */
    public QueryBuilder rightJoin (String join) {
        this.join = process(join, "RIGHT JOIN");
        return this;
    }
    
    
    
    /**
     * @param  join {String}
     * @return QueryBuilder
     * 
     * The purpose of this method is to allow a developer to 
     * perform a simple left join. 
     */
    public QueryBuilder leftJoin (String join) {
        this.join = process(join, "LEFT JOIN");
        return this;
    }
    
    
    
    /**
     * @param  on {String}
     * @return QueryBuilder
     * 
     * The purpose of this simple method is to allow a developer 
     * to complete the join so to speak. 
     */
    public QueryBuilder on (String on) {
        if (isValidString(on)) {
            
            if (!on.replaceAll(" ", "").toUpperCase().contains("ON")) {
                on = " ON " + on;
            }
            
            join += on;
            joins.add(join);
            join = "";
        }
        
        return this;
    }
    
    
    
    /**
     * @param  column1 {String}
     * @param  column2 {String}
     * @return QueryBuilder 
     * 
     * The purpose of this method is to mimic the one above, 
     * but allowing the developer to provide two strings instead 
     * of one long one. 
     */
    public QueryBuilder on (String column1, String column2) {
        if (!isValidString(column1) || !isValidString(column2)) {
            return this;
        }
        
        column1 = column1.replaceAll("=", "");
        column2 = column2.replaceAll("=", "");
        
        String on = " ON " + column1 + " = " + column2;
        join += on;
        joins.add(join);
        join = "";
    
        return this;
    }
    
    
    /**
     * @param  col {String}
     * @return QueryBuilder 
     * 
     * @todo   Allow for the ability to toggle
     *         the default sorting order.
     * 
     * The purpose of this method is to allow 
     * a developer to just order by a specific 
     * column, disregarding the actual order.
     * At the moment it currently will order by 
     * ascending order, however there's the idea 
     * to possibly allow a developer to somehow 
     * toggle this value somewhere.
     */
    public QueryBuilder orderBy (String col) {
        orderBy = process(col, "ORDER BY");
        orderBy += " ASC ";
        return this;
    }
    
    
    
    /**
     * @param  col   {String}
     * @param  order {String}
     * @return QueryBuilder 
     * 
     * The purpose of this method is to simply allow the 
     * developer to state how they'd like to order the query.
     */
    public QueryBuilder orderBy (String col, String order) {
        this.orderBy = process(col, "ORDER BY");
        
        if (isValidString(order)) {
            order = order.toUpperCase().replaceAll(" ", "");
            
            if ((!order.equals("ASC")) && (!order.equals("DESC"))) {
                order = "ASC";
            }
        } else {
            order = "ASC";
        }
        
        this.orderBy += order;
        
        return this;
    }
    
    
    
    /**
     * @param String {group}
     */
    public QueryBuilder groupBy (String group) {
        this.groupBy = process(group, "GROUP BY");
        return this;
    }
    
    
    
    /**
     * @return String 
     * 
     * The purpose of this method is to simply allow a developer 
     * to get the query string value from the current object.
     */
    public String build () {
        build = "";
        build += select + NEW_LINE;
        build += from + NEW_LINE;
        
        joins.forEach(j -> { build += j + NEW_LINE; });
        build += where + NEW_LINE;
        build += orderBy + groupBy;
        
        return build;
    }
}
